### DISCUSSION TITLE
Could you write a tauri command in main.rs to run window.eval in any labeled window?

### DISCUSSION THREAD

**Discussion by BubuInc** [tauri-apps/tauri#8792] [Q&A] [answered]
I'm new to tauri. I don't know rust, only javascript. I found that there isn't direct access to eval in `__TAURI__.window`, but only in main.rs as part of the window struct 
https://docs.rs/tauri/1.2.1/tauri/window/struct.Window.html#method.eval
Could you help me to write a tauri command to run window.eval in any labeled window, to use it in the frontend ?

I tried:
```rust
use tauri::Manager;

#[tauri::command]
fn eval(winLabel: &str, str: &str) {
    let app = tauri::Builder::default();
    let win = app.get_window(winLabel).unwrap();
    win.eval(str);
}
```
But it throws error: no method named `get_window` found for struct `tauri::Builder` in the current scope

I tried this too:
```rust
#[tauri::command]
fn eval(winLabel: &str, str: &str) {
    tauri::Builder::default().setup(|app| {
            let win = app.get_window(winLabel).unwrap();
            win.eval(str);
        });
}
```
But it throws error: expected `Result<(), Box<dyn, Error>>`, found `()`, after `|app|`

I added that Result after app but it throws the same error

**Comment by FabianLars** [comment, marked as answer]
The thing you're missing is this: https://tauri.app/v1/guides/features/command#accessing-an-apphandle-in-commands. In your case it would look something like this:
```rs
use tauri::Manager;

#[tauri::command]
fn eval(app: tauri::AppHandle, winLabel: &str, str: &str) {
    let win = app.get_window(winLabel).unwrap();
    win.eval(str);
}
```
You don't have to do anything special in the js side or anywhere to get the AppHandle into that command, tauri will inject it for you.

**Reply by BubuInc** [reply]
Good. It works. But how to get the eval result?
I tried:
```rust
let win = app.get_window(winLabel).unwrap();
let res = win.eval(str);
println!("{}", res);
``` 
But it throws error: `Result<(), tauri::Error>` cannot be formatted with the default formatter

**Reply by FabianLars** [reply]
There is no actual result if you're trying to get anything back from the frontend (eval does not return anything). If you want to print the execution error (if any) to the terminal, `println!("{:?}", res);` should work.

**Reply by jamesa08** [reply]
Are there any plans to make eval return a result, serialized in JSON?

**Reply by FabianLars** [reply]
It's not on our todo list as far as i know but that doesn't mean it won't be added.
